### PR TITLE
UMat usageFlags fixes opencv/opencv#19807

### PR DESCRIPTION
**Pull request by diablodale**
fixes opencv/opencv#19807

* Request code review of these changes before merge.
* I do not recommend merging this until after code review, then I try this on the 3.x branch, and then I make any needed changes. ***I have not done any coding or testing of this on the 3.x legacy branch.*** I request code review of this approach before I do anything in 3.x
* Clarified doxygen comments on some fields of `cv::UMat`
* Please consider my notes and caveats at https://github.com/opencv/opencv/issues/19807#issuecomment-810645975

I have been using this in my own code for one month with opencl on/off and with two GPUs (intel and nvidia).
Passes all core unit tests when applied to the 4.5.2 branch.
No known issues.

```
[----------] Global test environment tear-down
[ SKIPSTAT ] 36 tests skipped
[ SKIPSTAT ] TAG='mem_6gb' skip 1 tests
[ SKIPSTAT ] TAG='skip_bigdata' skip 1 tests
[ SKIPSTAT ] TAG='skip_other' skip 34 tests
[==========] 11621 tests from 250 test cases ran. (764388 ms total)
[  PASSED  ] 11621 tests.
  YOU HAVE 19 DISABLED TESTS
```

### Pull Request Readiness Checklist

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [X] The PR is proposed to proper branch
- [X] There is reference to original bug report and related work
- [X] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by diablodale** [comment]
I don't see any failures in the build related to my code changes.
Instead, the fails appear related to Flann code and to the test infrastructure itself.
If you think otherwise, please point me to the specific test case failure that is related to my code.

**Comment by mshabunin** [comment]
@diablodale , only _default_ check is required for now, so other issues can be ignored.

Could you please also add regression test(s) for this issue? Did you try to [compare perf results](https://github.com/opencv/opencv/wiki/HowToUsePerfTests#comparing-results-with-summarypy) before and after the patch?

**Comment by diablodale** [comment]
### New test cases

Yes, I will add regression tests for this issue. It will be in my next forcepush (which might include any needed changes for 3.x).

### Performance

This is a concern of mine that I discuss at the top of the original issue https://github.com/opencv/opencv/issues/19807
I believe we need to announce the fix in the next release. Something like

> Fixed a bug in our OpenCL memory management. Will not affect most users since USAGE_DEFAULT has not changed. If you use OpenCL in your OpenCV app and a usage type other than USAGE_DEFAULT, we recommend you test your app for any performance changes. If the changes are negative, then we recommend changing your code to USAGE_DEFAULT since that was the actual usage before this fix.

The specific use case of non-DEFAULT very much depends on the cpu, gpu, gpu's core and opencl driver, and the app code itself. This combination has interactions of PCIe buses, caches, memory timing, driver optimizations, opencl implementation, etc. The performance on a computer with a discrete NVidia 1080 is different than an integrated Intel HD. The NVidia almost always has to move memory across the PCIe bus. The Intel might be able to use the same RAM memory as the CPU. On both those cases, the three different USAGE settings will have different interactions and likely some measurable difference in performance if run millions of times. https://developer.nvidia.com/blog/how-optimize-data-transfers-cuda-cc/ is such a topic where NVidia writes

> You should not over-allocate [host] pinned memory....How much is too much is difficult to tell in advance, so as with all optimizations, test your applications and the systems they run on for optimal performance parameters.

I will write some test cases that iterate millions of times and include a report. It might tell us a relative difference, on my specific computer, specific GPUs, specific drivers, for a specific codepath in an app.
I have no problem doing this. 👍🙂
And...I don't think we will have any meaningful action with any result I get.

In my app that processes 5 streams of visual data at 30fps, I saw no consistent universal measurable perf changes. I thought I saw one case of measurable improvement, but it wasn't consistent.

### What data do we need from a performance test case?

What data do you think we need to have? Specifically.

What exact test case(s) do I need to write?

I can imagine 3 groups; one group for each usage type: `USAGE_DEFAULT`, `USAGE_ALLOCATE_HOST_MEMORY`, and `USAGE_ALLOCATE_DEVICE_MEMORY`

Then within a group, to do _something_ that is tests the allocation and transfer of memory...and doesn't have gpu computation work that would be magnitudes more costly -- that cost would overwhelm the transfer metrics. For example, it would probably be bad to run a DNN model. Perhaps `UMat::ones()*UMat::ones()` operation? 🤷‍♀️

Do I need to manipulate the OpenCV cache pool/allocator? Settings like `OPENCV_OPENCL_HOST_PTR_BUFFERPOOL_LIMIT` or `OPENCV_OPENCL_BUFFER_FORCE_MAPPING`, etc. ? 🤷‍♂️

**Comment by mshabunin** [comment]
No, it is not necessary to write new performance tests, I meant running existing test suite (at least for _core_, _imgproc_ modules): https://github.com/opencv/opencv/wiki/HowToUsePerfTests
Running...
```.sh
export OPENCV_TEST_DATA_PATH=...
# without patch
mkdir -p before
python3 opencv/modules/ts/misc/run.py -t core,imgproc -w before build
# with patch
mkdir -p after
python3 opencv/modules/ts/misc/run.py -t core,imgproc -w after build
```
...and comparing results with summary.py script
```.sh
for m in core imgproc ; do
  python3 opencv/modules/ts/misc/summary.py -o html before/*${m}*.xml after/*${m}*.xml > report_${m}.html
done
```
The idea is to check whether this patch makes _significant_ performance difference (faster or slower by >30%) on iGPU and dGPU (if possible). I don't think advanced settings should be changed.

We even have perf test for usage flags: https://github.com/opencv/opencv/blob/master/modules/core/perf/opencl/perf_usage_flags.cpp

#### Note
There can be cases when performance results are not stable, you will see lines like this in the test log (stdev is >3-5%):
```
[ PERFSTAT ]    (samples=100   mean=2.02   median=1.44   min=0.34   stddev=1.68 (83.4%))
```
If you experience these issues, it is recommended to prepare machine prior to testing: close all other applications, fixate CPU/GPU frequiencies, disable turbo boost, etc..

**Comment by diablodale** [comment]
Not seeing actionable data in the reports. Suggestions?

[report1.zip](https://github.com/opencv/opencv/files/6424495/report1.zip)

**Comment by diablodale** [comment]
Same chart and pivot table for the `imgproc module`...

Overall 2.1% faster
image

If I filter for only the "OCL" test cases...
overall 3.6% faster
image


Row Labels | Average   of %change, negative is slower
-- | --
OCL | 25.3%
OCL__3vs4_Fixture | -0.5%
OCL_AccumulateFixture | -3.6%
OCL_AccumulateProductFixture | 1.1%
OCL_AccumulateSquareFixture | -5.8%
OCL_AccumulateWeightedFixture | -0.6%
OCL_BilateralFixture | 1.2%
OCL_BlendLinearFixture | -2.5%
OCL_BlurFixture | 11.9%
OCL_BuildPyramidFixture | 0.9%
OCL_CalcBackProjFixture | 34.6%
OCL_CalcHistFixture | 29.4%
OCL_CannyFixture | 0.8%
OCL_CLAHEFixture | -4.1%
OCL_CopyMakeBorderFixture | 9.2%
OCL_CornerHarrisFixture | -0.6%
OCL_CornerMinEigenValFixture | 1.1%
OCL_CV_TM_CCORR_NORMEDFixture | 5.6%
OCL_CV_TM_CCORRFixture | 3.9%
OCL_CvtColorFixture | -0.3%
OCL_DilateFixture | 2.4%
OCL_EqualizeHistFixture | 27.0%
OCL_ErodeFixture | 2.6%
OCL_Filter2DFixture | 24.4%
OCL_GaussianBlurFixture | -16.7%
OCL_GoodFeaturesToTrackFixture | 1.7%
OCL_HoughLinesFixture | -1.6%
OCL_HoughLinesPFixture | 27.5%
OCL_ImgSize_TmplSize_Method_MatType | 2.5%
OCL_IntegralFixture | -0.7%
OCL_LaplacianFixture | 13.7%
OCL_MedianBlurFixture | 0.8%
OCL_MomentsFixture | -0.9%
OCL_MorphologyExFixture | 3.2%
OCL_PreCornerDetectFixture | 0.6%
OCL_PyrDownFixture | 11.8%
OCL_PyrUpFixture | 2.0%
OCL_RemapFixture | 1.9%
OCL_ResizeAreaFixture | 2.1%
OCL_ResizeFixture | 0.5%
OCL_ResizeLinearExactFixture | 0.0%
OCL_ScharrFixture | -11.1%
OCL_SepFilter2D | 20.5%
OCL_SobelFixture | -9.0%
OCL_SqrBoxFilterFixture | 9.7%
OCL_ThreshFixture | 3.3%
OCL_WarpAffineFixture | 6.4%
OCL_WarpPerspectiveFixture | 3.2%
Grand Total | 3.6%

**Comment by diablodale** [comment]
I don't have confidence in using the bulk test cases for this. Perhaps I am cautious.
The variability is too high. Today I ran the same tests again. Combined before and after the fix.
I got yet again some better some worse.

I generated a report to compare the before code (official 4.5.2) to *itself*. I compiled and ran yesterday, compiled and ran today. Generated the report with those two datasets. I got similar results...some better...some worse. And the variation was large. Here is the [report_core.zip](https://github.com/opencv/opencv/files/6429049/report_core.zip) and two lines from it

```
norm_mask::Size_MatType_NormType::(1920x1080, 32FC1, NORM_L2)     64% faster
CUDA_GpuMat_CopyToMasked::Sz_Depth_Cn::(1920x1080, CV_64F, Gray) 49% slower
```

I think the only thing I can learn from these bulk reports, is if all OCL tests were slower. If I saw a clear consistent pattern like that, then I would say...we have a code problem. But since the sample data varies by 50% with the same code run on two different days...I don't think much more can be learned.

In contrast, a focused test that runs much more than 13 (a very common # of samples the test harness runs) samples and over a longer period of time might give us more meaningful information. Yes, the caches in the hardware (cpu,gpu) will be engaged. But the traffic across the PCIe (or not for integrated) remains as well as cpu and gpu memory contention.

I'm going to write an additional perf test case that runs for a long time to see if I can discern anything. The existing UsageFlags perf test case does show performance differences when the different combinations of memory types are used. I want to make those clearer.

**Comment by mshabunin** [comment]
@diablodale , perf tests have parameters ` --perf_min_samples=10` and `--perf_force_samples=100` to set boundaries for number of samples, test usually stops either when the timeout (`--perf_time_limit=3.0`) is reached either when result deviation is less than defined (`--perf_max_deviation=1.0`). Also option `--gtest_filter=*OCL*` should enable only OpenCL tests. You can pass these parameters to run.py wrapper script or directly to test executables.

**Comment by diablodale** [comment]
Hi. The "default" CI pipeline isn't reporting results. It has been hours since I force pushed and it still reports "waiting for status..." here.

**Comment by diablodale** [comment]
I can not discern from the performance reports any significant and consistent improvements or slowdowns associated with the code fix. Same code run on two different days often has variations of 45% or more (xfactor 0.39, 0.52, etc.). However, it is a very small number of test cases that have these large variations. I believe they are outliers and test noise.

All charts include data that are...

* opencl enabled and my discrete nvidia rtx2070super selected using the envvar
* release x64 windows build
* windows 10
* thousands of test cases, therefore labeling test cases on an axis is usually not possible
* all data has equal weighting; therefore when parameterized test cases run many combinations of parameters (e.g .matrix sizes) that causes the test case to have a larger contribution to everything -- likely distorting some analysis -- please remember this
* Use ***microsecond*** resolution to get non-zero data due to roundoff. Yet...changes in the microsecond range are difficult to distinguish from test noise and generates xfactors that are likely invalid. For example
  ```
  // I doubt this is actually 50% faster and is instead test noise
  bitwise_xor::Size_MatType::(127x61, 8UC1)   before=0.3µs   after=0.2µs   xfactor=1.5
  ```
* report was created with `python3 ../modules/ts/misc/summary.py --units us -o html module_before.xml module_after.xml > fullus.html`. 
* this `fullus.html` report imported into Excel and charts generated

### Core

* 4287 data points (due to parameters) from 153 test cases within 121 test groups
* data point average is xfactor 1.015379053 (tiny improvement -- likely noise)
* test case average is xfactor 1.002387016 (tiny improvement -- likely noise)
* test group average is xfactor 1.002512992 (tiny improvement -- likely noise)

The super majority of test cases is located between 0.7 - 1.3 xfactor
image

All test cases sorted by xfactor from most improvement to least improvement. Y axis is the xfactor for the infinitely thin "bar" in the chart. It shows a "long tail" distribution on each side.
image

[Box and whisker chart](https://support.microsoft.com/en-us/office/create-a-box-and-whisker-chart-62f4219f-db4b-4754-aca8-4743f6190f0d) sorted by xfactor of individual test cases from most to least while also grouped by test group. This chart suggests that these extreme improvements are outliers and likely test noise. The more consistent test results are the shaded bars that are usually around 1.0.
image

Box and whisker chart reverse sorted. This suggests that these extreme declines are also test noise.
image
 
### Imgproc

* 5364 data points (due to parameters) from 147 test cases within 113 test groups
* data point average is xfactor 1.014306488 (tiny improvement -- likely noise)
* test case average is xfactor 1.011526064 (tiny improvement -- likely noise)
* test group average is xfactor 1.010814858 (tiny improvement -- likely noise)

The super majority of test cases is located between 0.7 - 1.3 xfactor
image

All test cases sorted by xfactor from most improvement to least improvement. Y axis is the xfactor for the infinitely thin "bar" in the chart. It shows a "long tail" distribution on each side.
image


Box and whisker sorted by xfactor of individual test cases from most to least while also grouped by test group. The imgproc module has more outliers, though this still does not concern me given the other research.
image

Box and whisker chart reverse sorted.
image

**Comment by diablodale** [comment]
@mshabunin do you have any comments regarding the above perf test results?
If no comments, then i will shift my focus to examine if something quick can be done to the 3.x branch.

File: modules/core/perf/opencl/perf_usage_flags.cpp [hunk 1]
```diff
@@ -12,25 +12,33 @@
...
-
-OCL_PERF_TEST_P(UsageFlagsBoolFixture, UsageFlags_AllocHostMem, ::testing::Combine(OCL_TEST_SIZES, Bool()))
+typedef TestBaseWithParam<tuple<cv::Size, UMatUsageFlags, UMatUsageFlags, UMatUsageFlags>> SizeUsageFlagsFixture;
+
+OCL_PERF_TEST_P(SizeUsageFlagsFixture, UsageFlags_AllocMem,
+    ::testing::Combine(
+        OCL_TEST_SIZES,
+        testing::Values(USAGE_DEFAULT, USAGE_ALLOCATE_HOST_MEMORY, USAGE_ALLOCATE_DEVICE_MEMORY), // USAGE_ALLOCATE_SHARED_MEMORY
+        testing::Values(USAGE_DEFAULT, USAGE_ALLOCATE_HOST_MEMORY, USAGE_ALLOCATE_DEVICE_MEMORY), // USAGE_ALLOCATE_SHARED_MEMORY
+        testing::Values(USAGE_DEFAULT, USAGE_ALLOCATE_HOST_MEMORY, USAGE_ALLOCATE_DEVICE_MEMORY) // USAGE_ALLOCATE_SHARED_MEMORY
```
Comment (alalek) [inline modules/core/perf/opencl/perf_usage_flags.cpp:22]: These tests are very long (>50% of whole opencv_perf_core):

> `108 tests from OCL_SizeUsageFlagsFixture_UsageFlags_AllocMem (104914 ms total)`

Perhaps we should replace OCL_TEST_SIZES with `::testing::Values(sz1080p)`

and revert this change:

> `OCL_TEST_CYCLE_MULTIRUN(500)`

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

Please take a look on comments below.

File: modules/core/include/opencv2/core/mat.hpp [hunk 2]
```diff
@@ -2555,27 +2555,38 @@ class CV_EXPORTS UMat
...
     UMatData* u;
 
-    // offset of the submatrix (or 0)
+    //! offset of the submatrix (or 0)
     size_t offset;
 
+    //! dimensional size of the matrix; accessible in various formats
     MatSize size;
+
+    //! number of bytes each matrix row occupies
```
[the remark this answers is not in the document]
Comment (diablodale) [reply modules/core/include/opencv2/core/mat.hpp:2589]: will be in next force push.

File: modules/ts/misc/table_formatter.py [hunk 3]
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env python
 
 from __future__ import print_function
-import sys, re, os.path, cgi, stat, math
+import sys, re, os.path, html, stat, math
```
[the remark this answers is not in the document]
Comment (diablodale) [reply modules/ts/misc/table_formatter.py:4]: this code will not be part of final PR. This is only a quick hack to get it working in python3.8 and newer.
see https://github.com/opencv/opencv/issues/20032

File: modules/core/perf/opencl/perf_usage_flags.cpp (hunk 1 above)
Comment (diablodale) [reply modules/core/perf/opencl/perf_usage_flags.cpp:22]: As I wrote in the OP...this PR is NOT NOT NOT ready to be merged. I have been seeking code and performance review. I have those two now...so I will start looking at 3.x and if I can readily backport this.

These long perf tests will not be part of the final PR. They are only in the current changeset so that I can show you the tests that I run to generate the above performance reports, and to have broader coverage on more platforms than I have locally. They will be used in my effort to backport to 3.x (if that is readily possible).

I will definitely remove the OCL_TEST_CYCLE_MULTIRUN() in final code. AT the same time, I will evaluate if I can continue to use `OCL_TEST_SIZES`. We might even consider making this perf test disabled.

File: modules/core/src/umatrix.cpp [hunk 4]
```diff
@@ -610,16 +609,18 @@ UMat Mat::getUMat(AccessFlag accessFlags, UMatUsageFlags usageFlags) const
...
     int i;
     CV_Assert(0 <= d && d <= CV_MAX_DIM && _sizes);
     _type = CV_MAT_TYPE(_type);
 
-    if( u && (d == dims || (d == 1 && dims <= 2)) && _type == type() )
+    // if param value is USAGE_DEFAULT by implicit default param value -or- explicit value
+    // ...then don't change the existing usageFlags
+    // it is not possible to change usage from non-default to USAGE_DEFAULT through create()
+    // ...instead must construct UMat()
+    if (_usageFlags == cv::USAGE_DEFAULT) _usageFlags = usageFlags;
```
[the remark this answers is not in the document]
Comment (diablodale) [reply modules/core/src/umatrix.cpp:620]: I am not clear on what you are asking. Do you want me to write it like...???
```c++
if (_usageFlags == cv::USAGE_DEFAULT) {
    _usageFlags = usageFlags;
}
```

[the remark this answers is not in the document]
Comment (alalek) [reply modules/core/src/umatrix.cpp:620]: yes

[the remark this answers is not in the document]
Comment (diablodale) [reply modules/core/src/umatrix.cpp:620]: thx for clarifying. it will be in next force push

**Comment by diablodale** [comment]
@alalek, this PR can not be applied to the 3.x branch. That branch has no support for move-semantics. Therefore, there are no methods and operator= that use move-semantics. This PR has changes which fix issues in those move methods and if you try to apply, it will fail because those methods don't exist.

What approach do you recommend?

**Comment by alalek** [comment]
Please continue working on the "master" branch.
I will cherry-pick changes after merge of this PR if needed (3.4 branch will get low profile support only soon).

**Comment by diablodale** [comment]
Ok. I will now...

* remove the MULTIRUN(500) and casually experiment with its run time so it is not excessive while also considering if that specific perf test should be disabled
* remove the python compatibility workaround
* flatten, rebase, and force push again to master

If you potentially cherry-pick to 3.4 branch... I want to caution that...

* I have only done a visual code review of this PR's changes on 3.4.
* I think a subset of this PR will work. The subset would be all the changes except for the move-semantics methods/operators 
* I have not compiled these changes on 3.4 branch. They might need very minor adjustments when cherry-picked.
* I have not done unit, functional, or performance testing on 3.4 branch. _I do not recommend applying this PR to 3.4 without all this testing_

File: modules/core/perf/opencl/perf_usage_flags.cpp (hunk 1 above)
Comment (diablodale) [reply modules/core/perf/opencl/perf_usage_flags.cpp:22]: updated PR no longer loops 500. On my test machine, the test you write above is smaller than other perf tests, e.g. the Dft test is 3 times longer.

```
144 tests from OCL_DftFixture_Dft (39435 ms total)
...
108 tests from OCL_SizeUsageFlagsFixture_UsageFlags_AllocMem (13932 ms total)
```

Is this length ok?
If you still want to remove sizes, then I recommend we keep only `OCL_SIZE_1` (which is szVGA) and `OCL_SIZE_4` (which is sz2160p) so that we have the smallest, and largest. That will probably half again the run time.

Comment (diablodale) [reply modules/core/perf/opencl/perf_usage_flags.cpp:22]: the default CI on linux x64 opencl and windows opencl shows the perf test runs in single-digit seconds, or in less than one second. Dft is still larger on both.

I think we can keep the perf test as it is now. Agree?
If yes, then I have no further work to do and, from my eval, this PR is ready to merge into master.

**Review by alalek — APPROVED** [review]
Looks good! Thank you :+1:
